pysal/contrib/viz/mapping.py

```diff
@@ -610,11 +611,23 @@ def geoplot(db, col=None, palette='BuGn', classi='Quantiles',
                   Additional named vaues to be passed to the classifier of choice.
     '''
     if col:
+        if hasattr(palette, 'number') and 'k' not in kwargs:
+            # infer number of classes from available colors
+            kwargs['k'] = palette.number
+        elif hasattr(palette, 'number') and 'k' in kwargs:
+            if kwargs['k'] != palette.number:
+                warn('explicit "k" argument differs from the number of colors in '
+                'the palette. Defaulting to the number of colors in the palette.', 
+                stacklevel=1)
+                kwargs['k'] = palette.number
```

Comment:
I see, I'd missed l.19 et al. :-) Right. Only thing I'm thinking is whether we'd want to just fail with a sensible traceback message instead. If you put the effort of setting `k` and a `palette` specifically, probably getting only one of the two is not you desired outcome.

I agree it'd be nice to dynamically "edit" the number of colors in a palette, but not sure how straightforward it is...